File: training/environment/game_2048.py
import gym
from gym import spaces
import numpy as np
import random
from typing import Tuple, Dict, Any, Optional, List

class IllegalMove(Exception):
    """불가능한 움직임을 시도할 때 발생하는 예외"""
    pass

def stack(flat: np.ndarray, layers: int = 16) -> np.ndarray:
    """
    Convert an [4, 4] representation into [4, 4, layers] with one layer for each value.
    
    Args:
        flat: 4x4 보드 배열
        layers: 레이어 수 (기본값: 16)
    
    Returns:
        4x4x16 원-핫 인코딩된 배열
    """
    representation = 2 ** (np.arange(layers, dtype=int) + 1)
    layered = np.repeat(flat[:,:,np.newaxis], layers, axis=-1)
    layered = np.where(layered == representation, 1, 0)
    return layered.astype(np.float32)

class Game2048Env(gym.Env):
    """
    2048 게임 Gym 환경 (액션 마스킹 지원)
    
    Action Space: Discrete(4) - 0: 위, 1: 오른쪽, 2: 아래, 3: 왼쪽
    Observation Space: Box(0, 1, (4, 4, 16), dtype=np.float32) - 원-핫 인코딩
    """
    
    metadata = {'render.modes': ['human', 'ansi']}
    
    def __init__(self, size: int = 4, max_steps: int = 10000):
        super(Game2048Env, self).__init__()
        
        if size <= 0:
            raise ValueError("Size must be positive")
        if max_steps <= 0:
            raise ValueError("max_steps must be positive")
        
        self.size = size
        self.squares = self.size * self.size
        self.max_steps = max_steps
        
        # 액션/관찰 공간
        self.action_space = spaces.Discrete(4)
        self.observation_space = spaces.Box(
            low=0, high=1, 
            shape=(self.size, self.size, self.squares), 
            dtype=np.float32
        )
        
        # 게임 상태
        self.board: Optional[np.ndarray] = None
        self.score: int = 0
        self.steps: int = 0
        self._game_initialized: bool = False
        
        # 보상 설정
        # self.reward_range = (0.0, float(2**self.squares))
    
    def reset(self) -> np.ndarray:
        """환경 초기화"""
        self.board = np.zeros((self.size, self.size), dtype=int)
        self.score = 0
        self.steps = 0
        self._game_initialized = True
        
        # 초기 타일 2개 추가
        self._add_random_tile()
        self._add_random_tile()
        
        return self._get_observation()
    
    def step(self, action: int) -> Tuple[np.ndarray, float, bool, Dict[str, Any]]:
        """한 스텝 실행"""
        if not self._game_initialized:
            raise RuntimeError("Environment not initialized. Call reset() first.")
        
        if not self.action_space.contains(action):
            raise ValueError(f"Invalid action: {action}")
        
        self.steps += 1
        info = {'illegal_move': False}
        done = False
        
        try:
            move_score = self._move(action)
            self.score += move_score
            self._add_random_tile()
            done = self._is_game_over()
            reward = self._calculate_reward(move_score)
            
        except IllegalMove:
            info['illegal_move'] = True
            reward = 0.0
        
        # 게임 종료 조건 (스텝 수 제한)
        if not done and self.steps >= self.max_steps:
            done = True
        
        info.update({
            'score': int(self.score),
            'highest': int(np.max(self.board)),
            'empty_cells': len(self._get_empty_cells()),
            'steps': self.steps,
            'valid_actions': self.get_valid_actions(),
            'move_score': move_score if 'move_score' in locals() else 0
        })
        
        return self._get_observation(), reward, done, info

    def render(self, mode: str = 'human') -> Optional[str]:
        """게임 상태 시각화"""
        if not self._game_initialized:
            print("Game not initialized. Call reset() first.")
            return None
            
        output = self._render_ansi()
        if mode == 'human':
            print(output)
            return None
        elif mode == 'ansi':
            return output
        else:
            raise NotImplementedError(f"Render mode '{mode}' not supported")

    def _render_ansi(self) -> str:
        """ANSI 형태로 보드 렌더링"""
        if not self._game_initialized:
            return "Game not initialized. Call reset() first.\n"
        
        s = f"Score: {self.score} | Steps: {self.steps}\n"
        s += f"Highest: {int(np.max(self.board))} | Empty: {len(self._get_empty_cells())}\n"
        s += "-" * 25 + "\n"
        
        for row in self.board:
            s += "|"
            for cell in row:
                if cell == 0:
                    s += "    ."
                else:
                    s += f"{cell:5d}"
            s += " |\n"
        
        s += "-" * 25 + "\n"
        return s

    def _calculate_reward(self, move_score: float) -> float:
        """휴리스틱 보상 함수 (기본형)"""
        # 가중치
        W_MERGE = 1.0
        W_EMPTY = 2.7

        # 1. 합병 점수
        merge_reward = np.log2(max(move_score, 1)) if move_score > 0 else 0

        # 2. 빈 타일 보상
        empty_cells = len(self._get_empty_cells())
        empty_reward = np.log(max(empty_cells, 1)) if empty_cells > 0 else 0

        total_reward = (
            W_MERGE * merge_reward +
            W_EMPTY * empty_reward
        )
        
        return float(total_reward)
    
    def _add_random_tile(self) -> bool:
        """빈 자리에 랜덤 타일 추가"""
        empty_cells = self._get_empty_cells()
        if not empty_cells:
            return False
        
        i, j = random.choice(empty_cells)
        self.board[i][j] = 2 if random.random() < 0.9 else 4
        return True

    def _get_empty_cells(self) -> List[Tuple[int, int]]:
        """빈 셀들의 좌표 반환"""
        return [(i, j) for i in range(self.size) for j in range(self.size) if self.board[i][j] == 0]
    
    def _move(self, direction: int) -> float:
        """지정된 방향으로 보드 이동"""
        changed = False
        move_score = 0
        dir_div_two = int(direction / 2)
        dir_mod_two = int(direction % 2)
        shift_direction = dir_mod_two ^ dir_div_two
        
        if dir_mod_two == 0:
            # 수직 이동
            for j in range(self.size):
                col = [self.board[i][j] for i in range(self.size)]
                (new_col, score) = self._shift(col, shift_direction)
                move_score += score
                for i in range(self.size):
                    if self.board[i][j] != new_col[i]:
                        changed = True
                    self.board[i][j] = new_col[i]
        else:
            # 수평 이동
            for i in range(self.size):
                row = [self.board[i][j] for j in range(self.size)]
                (new_row, score) = self._shift(row, shift_direction)
                move_score += score
                for j in range(self.size):
                    if self.board[i][j] != new_row[j]:
                        changed = True
                    self.board[i][j] = new_row[j]
        
        if not changed:
            raise IllegalMove(f"Move {direction} is not valid")
        
        return move_score
    
    def _shift(self, row: List[int], direction: int) -> Tuple[List[int], int]:
        """한 행을 이동하며 병합"""
        shifted_row = [i for i in row if i != 0]
        
        if direction:
            shifted_row.reverse()
        
        (combined_row, move_score) = self._combine(shifted_row)
        
        while len(combined_row) < self.size:
            combined_row.append(0)
        
        if direction:
            combined_row.reverse()
        
        return (combined_row, move_score)
    
    def _combine(self, shifted_row: List[int]) -> Tuple[List[int], int]:
        """같은 타일을 병합"""
        if not shifted_row:
            return ([0] * self.size, 0)
        
        move_score = 0
        combined_row = []
        i = 0
        
        while i < len(shifted_row):
            if (i + 1 < len(shifted_row) and 
                shifted_row[i] == shifted_row[i + 1] and 
                shifted_row[i] != 0):
                # 같은 타일 병합
                merged_value = 2 * shifted_row[i]
                combined_row.append(merged_value)
                move_score += merged_value
                i += 2
            else:
                combined_row.append(shifted_row[i])
                i += 1
        
        return (combined_row, move_score)
    
    def _is_game_over(self) -> bool:
        """게임 종료 여부 확인"""
        if self._get_empty_cells():
            return False
        
        # 인접한 같은 숫자 확인
        for i in range(self.size):
            for j in range(self.size):
                current = self.board[i][j]
                if j < self.size - 1 and current == self.board[i][j + 1]:
                    return False
                if i < self.size - 1 and current == self.board[i + 1][j]:
                    return False
        
        return True

    def _get_observation(self) -> np.ndarray:
        """현재 상태를 관찰 공간에 맞게 변환"""
        if not self._game_initialized:
            raise RuntimeError("Game not initialized. Call reset() first.")
        
        return stack(self.board.astype(np.float32), layers=self.squares)

    def get_board(self) -> np.ndarray:
        """현재 보드 상태 반환"""
        if not self._game_initialized:
            raise RuntimeError("Game not initialized. Call reset() first.")
        return self.board.copy()

    def get_valid_actions(self) -> List[int]:
        """현재 상태에서 유효한 액션들을 반환"""
        if not self._game_initialized:
            return []
        
        valid_actions = []
        for action in range(4):
            if self._is_valid_move(action):
                valid_actions.append(action)
        
        return valid_actions

    def _is_valid_move(self, action: int) -> bool:
        """특정 액션이 유효한지 확인"""
        original_board = self.board.copy()
        original_score = self.score
        
        try:
            self._move(action)
            return True
        except IllegalMove:
            return False
        finally:
            self.board = original_board
            self.score = original_score

    def get_action_mask(self) -> np.ndarray:
        """액션 마스크를 numpy 배열로 반환"""
        if not self._game_initialized:
            return np.zeros(4, dtype=bool)
        
        mask = np.zeros(4, dtype=bool)
        valid_actions = self.get_valid_actions()
        
        for action in valid_actions:
            mask[action] = True
        
        return mask

File: src/lib/game2048.ts
import { GameState, GameInfo, GameAction } from '@/types/game';

export class Game2048 {
  private board: number[][] = [];
  private score: number = 0;
  private steps: number = 0;
  private gameOver: boolean = false;
  private readonly size: number = 4;

  constructor() {
    this.reset();
  }

  reset(): GameState {
    this.board = Array(this.size).fill(null).map(() => Array(this.size).fill(0));
    this.score = 0;
    this.steps = 0;
    this.gameOver = false;
    
    // 초기 타일 2개 추가
    this.addRandomTile();
    this.addRandomTile();
    
    return this.getState();
  }

  step(action: GameAction): { state: GameState; reward: number; done: boolean; info: GameInfo } {
    if (this.gameOver) {
      return {
        state: this.getState(),
        reward: 0,
        done: true,
        info: this.getInfo()
      };
    }

    this.steps++;
    let reward = 0;
    let illegalMove = false;

    try {
      const moveScore = this.move(action);
      this.score += moveScore;
      const emptyCells = this.getEmptyCells();
      reward = this.calculateReward(moveScore, emptyCells.length); // 🔥 개선된 보상 함수 사용
      
      // 새 타일 추가
      if (!this.addRandomTile(emptyCells)) {
        // 타일 추가 실패 - 보드가 가득 참
        this.gameOver = this.isGameOver();
      } else {
        // 게임 종료 확인
        this.gameOver = this.isGameOver();
      }
    } catch {
      illegalMove = true;
      reward = -10; // illegal move에 대한 페널티
      // 🔥 액션 마스킹이 적용되면 이 부분은 실행되지 않아야 함
      console.warn(`⚠️ Warning: Illegal move ${action} attempted! Action masking should prevent this.`);
    }

    return {
      state: this.getState(),
      reward,
      done: this.gameOver,
      info: {
        ...this.getInfo(),
        illegal_move: illegalMove
      }
    };
  }

  // 🔥 새로운 개선된 보상 함수 추가 (Python 환경과 동일)
  private calculateReward(moveScore: number): number {
    // 가중치 (Python 환경과 동일)
    const W_MERGE = 1.0;
    const W_EMPTY = 2.7;
    const W_MONO = 1.0;
    const W_SMOOTH = 0.1;

    // 1. 합병 점수
    const mergeReward = moveScore > 0 ? Math.log2(moveScore) : 0;

    // 2. 빈 타일 보상
    const emptyCells = this.getEmptyCells().length;
    const emptyReward = emptyCells > 0 ? Math.log(emptyCells) : 0;

    // 3. 단조성 보상
    const monoReward = this.calculateMonotonicity();

    // 4. 평탄성 보상
    const smoothReward = this.calculateSmoothness();

    const totalReward = (
      W_MERGE * mergeReward +
      W_EMPTY * emptyReward +
      W_MONO * monoReward +
      W_SMOOTH * smoothReward
    );
    
    return totalReward;
  }

  // 🔥 단조성 계산 함수 추가
  private calculateMonotonicity(): number {
    let monotonicityScore = 0;
    
    // 행 단조성
    for (let i = 0; i < this.size; i++) {
      const rowValues = this.board[i].filter(cell => cell !== 0);
      if (rowValues.length > 1) {
        const logVals = rowValues.map(val => Math.log2(val));
        const increasing = logVals.slice(1).reduce((sum, val, idx) => sum + (val - logVals[idx]), 0);
        const decreasing = logVals.slice(1).reduce((sum, val, idx) => sum + (logVals[idx] - val), 0);
        monotonicityScore += Math.max(increasing, decreasing);
      }
    }

    // 열 단조성
    for (let j = 0; j < this.size; j++) {
      const colValues = [];
      for (let i = 0; i < this.size; i++) {
        if (this.board[i][j] !== 0) {
          colValues.push(this.board[i][j]);
        }
      }
      if (colValues.length > 1) {
        const logVals = colValues.map(val => Math.log2(val));
        const increasing = logVals.slice(1).reduce((sum, val, idx) => sum + (val - logVals[idx]), 0);
        const decreasing = logVals.slice(1).reduce((sum, val, idx) => sum + (logVals[idx] - val), 0);
        monotonicityScore += Math.max(increasing, decreasing);
      }
    }
    
    return monotonicityScore;
  }

  // 🔥 평탄성 계산 함수 추가
  private calculateSmoothness(): number {
    let smoothnessScore = 0;
    
    // 수평 평탄성
    for (let i = 0; i < this.size; i++) {
      for (let j = 0; j < this.size - 1; j++) {
        if (this.board[i][j] !== 0 && this.board[i][j + 1] !== 0) {
          smoothnessScore -= Math.abs(Math.log2(this.board[i][j]) - Math.log2(this.board[i][j + 1]));
        }
      }
    }

    // 수직 평탄성
    for (let i = 0; i < this.size - 1; i++) {
      for (let j = 0; j < this.size; j++) {
        if (this.board[i][j] !== 0 && this.board[i + 1][j] !== 0) {
          smoothnessScore -= Math.abs(Math.log2(this.board[i][j]) - Math.log2(this.board[i + 1][j]));
        }
      }
    }
    
    return smoothnessScore;
  }

  private move(direction: GameAction): number {
    const { board, score, changed } = this.testMove(direction, this.board);

    if (!changed) {
      throw new Error("No valid moves available");
    }

    this.board = board;
    return score;
  }

  private testMove(direction: GameAction, board: number[][]): { board: number[][], score: number, changed: boolean } {
    let moveScore = 0;
    let changed = false;
    const newBoard = board.map(row => [...row]);

    const dirDivTwo = Math.floor(direction / 2);
    const dirModTwo = direction % 2;
    const shiftDirection = dirModTwo ^ dirDivTwo;

    if (dirModTwo === 0) {
      // Up or down, split into columns
      for (let y = 0; y < this.size; y++) {
        const oldColumn = [];
        for (let x = 0; x < this.size; x++) {
          oldColumn.push(newBoard[x][y]);
        }
        
        const [newColumn, score] = this.shift(oldColumn, shiftDirection);
        moveScore += score;
        
        if (!this.arraysEqual(oldColumn, newColumn)) {
          changed = true;
          for (let x = 0; x < this.size; x++) {
            newBoard[x][y] = newColumn[x];
          }
        }
      }
    } else {
      // Left or right, split into rows
      for (let x = 0; x < this.size; x++) {
        const oldRow = [...newBoard[x]];
        const [newRow, score] = this.shift(oldRow, shiftDirection);
        moveScore += score;
        
        if (!this.arraysEqual(oldRow, newRow)) {
          changed = true;
          newBoard[x] = newRow;
        }
      }
    }

    return { board: newBoard, score: moveScore, changed };
  }

  private shift(row: number[], direction: number): [number[], number] {
    // Shift all non-zero digits
    const shiftedRow = row.filter(cell => cell !== 0);
    
    // Reverse if shifting right
    if (direction) {
      shiftedRow.reverse();
    }
    
    const [combinedRow, moveScore] = this.combine(shiftedRow);
    
    // Reverse back if shifting right
    if (direction) {
      combinedRow.reverse();
    }
    
    return [combinedRow, moveScore];
  }

  private combine(shiftedRow: number[]): [number[], number] {
    let moveScore = 0;
    const combinedRow = Array(this.size).fill(0);
    let skip = false;
    let outputIndex = 0;
    
    for (let i = 0; i < shiftedRow.length - 1; i++) {
      if (skip) {
        skip = false;
        continue;
      }
      
      combinedRow[outputIndex] = shiftedRow[i];
      
      if (shiftedRow[i] === shiftedRow[i + 1]) {
        combinedRow[outputIndex] += shiftedRow[i + 1];
        moveScore += shiftedRow[i] + shiftedRow[i + 1];
        skip = true;
      }
      outputIndex++;
    }
    
    if (shiftedRow.length > 0 && !skip) {
      combinedRow[outputIndex] = shiftedRow[shiftedRow.length - 1];
    }
    
    return [combinedRow, moveScore];
  }

  private addRandomTile(): boolean {
    const emptyCells = this.getEmptyCells();
    if (emptyCells.length === 0) return false;
    
    const randomCell = emptyCells[Math.floor(Math.random() * emptyCells.length)];
    const [i, j] = randomCell;
    
    // 90% 확률로 2, 10% 확률로 4
    this.board[i][j] = Math.random() < 0.9 ? 2 : 4;
    return true;
  }

  private getEmptyCells(): [number, number][] {
    const emptyCells: [number, number][] = [];
    for (let i = 0; i < this.size; i++) {
      for (let j = 0; j < this.size; j++) {
        if (this.board[i][j] === 0) {
          emptyCells.push([i, j]);
        }
      }
    }
    return emptyCells;
  }

  private isGameOver(): boolean {
    // 빈 칸이 있으면 게임 계속
    if (this.getEmptyCells().length > 0) return false;
    
    // 인접한 같은 숫자가 있는지 확인
    for (let i = 0; i < this.size; i++) {
      for (let j = 0; j < this.size; j++) {
        const current = this.board[i][j];
        // 오른쪽 인접 셀 확인
        if (j < this.size - 1 && this.board[i][j + 1] === current) return false;
        // 아래쪽 인접 셀 확인
        if (i < this.size - 1 && this.board[i + 1][j] === current) return false;
      }
    }
    
    return true;
  }

  private arraysEqual(a: number[], b: number[]): boolean {
    return a.length === b.length && a.every((val, i) => val === b[i]);
  }

  getState(): GameState {
    return {
      board: this.board.map(row => [...row]),
      score: this.score,
      steps: this.steps,
      highest: Math.max(...this.board.flat()),
      isEmpty: this.getEmptyCells().length > 0,
      gameOver: this.gameOver
    };
  }

  private getInfo(): GameInfo {
    return {
      score: this.score,
      highest: Math.max(...this.board.flat()),
      empty_cells: this.getEmptyCells().length,
      steps: this.steps,
      illegal_move: false
    };
  }

  // 🔥 핵심 수정: 학습 환경과 동일한 layered 관찰 생성
  getObservation(): Float32Array {
    // (4, 4, 16) 형태의 layered observation 생성
    const observation = new Float32Array(4 * 4 * 16);
    
    for (let i = 0; i < 4; i++) {
      for (let j = 0; j < 4; j++) {
        const value = this.board[i][j];
        if (value > 0) {
          // 2^n 값을 n-1 인덱스로 변환 (2→0, 4→1, 8→2, ...)
          const layerIndex = Math.log2(value) - 1;
          if (layerIndex >= 0 && layerIndex < 16) {
            const index = i * 4 * 16 + j * 16 + layerIndex;
            observation[index] = 1.0;
          }
        }
      }
    }
    
    return observation;
  }

  // 🔥 유효한 액션인지 미리 확인하는 메서드 (성능 최적화)
  isValidAction(action: GameAction): boolean {
    const { changed } = this.testMove(action, this.board);
    return changed;
  }

  // 🔥 가능한 액션들 반환 - 액션 마스킹의 핵심
  getValidActions(): GameAction[] {
    const validActions: GameAction[] = [];
    for (let action = 0; action < 4; action++) {
      if (this.isValidAction(action as GameAction)) {
        validActions.push(action as GameAction);
      }
    }
    return validActions;
  }

  getBoard(): number[][] {
    return this.board.map(row => [...row]);
  }
}